R/content-tools.R

```diff
@@ -44,35 +42,62 @@ extract_tool_requests <- function(contents) {
   contents[is_tool_request]
 }
 
+tool_result_factory <- function(tool, arguments, id) {
+  function(result = NULL, error = NULL) {
+    if (S7_inherits(result, ContentToolResult)) {
+      result@id <- id
+      result@call_tool <- tool
+      result@call_args <- arguments
+      return(result)
+    }
+
+    ContentToolResult(
+      value = result,
+      error = error,
+      id = id,
+      call_tool = tool,
+      call_args = arguments
+    )
+  }
+}
+
 # Also need to handle edge cases: https://platform.openai.com/docs/guides/function-calling/edge-cases
-invoke_tool <- function(fun, arguments, id) {
-  if (is.null(fun)) {
-    return(ContentToolResult(id = id, error = "Unknown tool"))
+invoke_tool <- function(tool, arguments, id) {
+  tool_result <- tool_result_factory(tool, arguments, id)
+
+  if (is.null(tool) || is.null(tool@fun)) {
```

Comment:
Why would `tool@fun` be `NULL`?

Reply:
...I'm not sure. It was the test inside of `invoke_tool()`, so I kept it. But you can't have a `ToolDef` with null `@fun`, and even `$set_tools()` goes through `$register_tool()`, which has a type check, so I think we can drop the null `fun` test.